File: scminerViewer/tests/testthat/test-browser.R
# Tests for discover_studies() + build_browser(): multi-study root with
# the <studyID>/<studyID>.scminer.h5 layout written by prepare_study_data.

test_that("discover_studies returns empty data.frame for missing/empty dir", {
  d <- tempfile("empty_")
  on.exit(unlink(d, recursive = TRUE), add = TRUE)
  expect_equal(nrow(discover_studies(d)), 0L)
  dir.create(d)
  expect_equal(nrow(discover_studies(d)), 0L)
})

test_that("discover_studies finds all <studyID>/<studyID>.scminer.h5 bundles", {
  root <- tempfile("studies_")
  on.exit(unlink(root, recursive = TRUE), add = TRUE)

  # Build two studies into the same root with different IDs.
  s1 <- synthetic_study(seed = 1)
  s1$meta$studyID    <- "1111"
  s1$meta$studyAbbr  <- "one"
  s1$meta$shortTitle <- "One"
  prepare_study_data(out_dir = root, meta = s1$meta,
                     cells = s1$cells, clusters = s1$clusters,
                     genes = s1$genes,
                     expression  = s1$expression,
                     activity_tf = s1$activity_tf,
                     activity_sig = s1$activity_sig,
                     network_tf = s1$network_tf,
                     network_sig = s1$network_sig,
                     emit = "bundle")

  s2 <- synthetic_study(seed = 2)
  s2$meta$studyID    <- "2222"
  s2$meta$studyAbbr  <- "two"
  s2$meta$shortTitle <- "Two"
  prepare_study_data(out_dir = root, meta = s2$meta,
                     cells = s2$cells, clusters = s2$clusters,
                     genes = s2$genes,
                     expression  = s2$expression,
                     emit = "bundle")

  df <- discover_studies(root)
  expect_equal(nrow(df), 2L)
  expect_setequal(df$studyID,   c("1111", "2222"))
  expect_setequal(df$studyAbbr, c("one",  "two"))
  expect_true(all(file.exists(df$bundle_path)))
  expect_true(all(df$n_cells > 0))
})

test_that("build_browser constructs a shiny app from a multi-study root", {
  skip_if_not_installed("bslib")
  skip_if_not_installed("plotly")
  skip_if_not_installed("DT")
  skip_if_not_installed("visNetwork")

  root <- tempfile("studies_")
  on.exit(unlink(root, recursive = TRUE), add = TRUE)
  s <- synthetic_study(seed = 3)
  prepare_study_data(out_dir = root, meta = s$meta,
                     cells = s$cells, clusters = s$clusters,
                     genes = s$genes,
                     expression = s$expression,
                     emit = "bundle")

  app <- build_browser(root)
  expect_s3_class(app, "shiny.appobj")
})

File: scminer_viewer/tests/test_browser.py
"""Tests for the Python multi-study browser API."""

from __future__ import annotations

import shutil
import subprocess
from pathlib import Path

import pandas as pd
import pytest

from scminer_viewer import discover_studies, build_browser


def _rscript_available() -> bool:
    return shutil.which("Rscript") is not None


PROJECT_ROOT = Path(__file__).resolve().parents[2]


@pytest.fixture(scope="module")
def two_study_root(tmp_path_factory) -> Path:
    """Build a root with two distinct synthetic studies via R."""
    if not _rscript_available():
        pytest.skip("Rscript not on PATH; cannot build fixture studies.")
    root = tmp_path_factory.mktemp("multistudies")
    root_str = repr(str(root))
    rscript = f"""
    suppressPackageStartupMessages(library(scminerViewer))
    make_one <- function(sid, abbr, seed) {{
      set.seed(seed)
      n_cells <- 20; n_genes <- 10
      cellTypes <- c("A","B")
      cells <- data.frame(
        cellID = paste0("c", seq_len(n_cells)),
        cellType = sample(cellTypes, n_cells, replace = TRUE),
        cellGroup = sample(cellTypes, n_cells, replace = TRUE),
        coord1 = rnorm(n_cells), coord2 = rnorm(n_cells),
        stringsAsFactors = FALSE
      )
      cnt <- as.data.frame(table(cells$cellType), stringsAsFactors=FALSE)
      clusters <- data.frame(
        cellType = cnt$Var1, count = cnt$Freq,
        color = c("#aa0000","#0000aa")[seq_len(nrow(cnt))],
        stringsAsFactors = FALSE
      )
      genes <- paste0("g", sprintf("%03d", seq_len(n_genes)))
      M <- Matrix::rsparsematrix(n_genes, n_cells, density = 0.3,
                                  rand.x = function(n) abs(rnorm(n)))
      rownames(M) <- genes; colnames(M) <- cells$cellID
      net <- data.frame(
        source = sample(genes, 5, replace=TRUE),
        target = sample(genes, 5, replace=TRUE),
        cellType = sample(cellTypes, 5, replace=TRUE),
        mi = runif(5), pearson = runif(5,-1,1),
        spearman = runif(5,-1,1), rho = runif(5,-1,1),
        pvalue = runif(5), stringsAsFactors=FALSE
      )
      prepare_study_data(
        out_dir = {root_str},
        meta = list(studyID = sid, studyAbbr = abbr,
                    longTitle = paste("Study", sid),
                    shortTitle = abbr,
                    species = "Mus musculus", coordinate = "UMAP"),
        cells = cells, clusters = clusters, genes = genes,
        expression = M, activity_tf = M*2, activity_sig = M*3,
        network_tf = net, network_sig = net,
        default_genes = genes[1:2],
        emit = c("graph", "bundle")
      )
    }}
    make_one("1111", "one", 1)
    make_one("2222", "two", 2)
    """
    res = subprocess.run(
        ["Rscript", "-e", rscript],
        capture_output=True, text=True, cwd=PROJECT_ROOT,
    )
    if res.returncode != 0:
        pytest.skip(f"Could not build fixture studies via R:\n{res.stderr}")
    return root


def test_discover_studies_empty_dir(tmp_path):
    empty = tmp_path / "empty"
    df = discover_studies(empty)
    assert isinstance(df, pd.DataFrame)
    assert df.empty
    assert set(df.columns) >= {
        "studyID", "studyAbbr", "shortTitle", "longTitle",
        "species", "n_cells", "n_genes", "n_clusters",
        "bundle_path", "study_dir",
    }


def test_discover_studies_finds_all_bundles(two_study_root):
    df = discover_studies(two_study_root)
    assert len(df) == 2
    assert set(df["studyID"]) == {"1111", "2222"}
    assert set(df["studyAbbr"]) == {"one", "two"}
    assert (df["n_cells"] > 0).all()
    assert all(Path(p).exists() for p in df["bundle_path"])


def test_discover_studies_sorted_by_id(two_study_root):
    df = discover_studies(two_study_root)
    assert list(df["studyID"]) == sorted(df["studyID"])


def test_build_browser_constructs_app(two_study_root):
    from shiny import App
    app = build_browser(two_study_root)
    assert isinstance(app, App)


def test_build_browser_with_shard_dir(two_study_root):
    from shiny import App
    # In the typical case the shards are already co-located; here we
    # simply verify the parameter is plumbed through.
    app = build_browser(two_study_root, shard_dir=two_study_root)
    assert isinstance(app, App)
